tutorials/pde/2_advection_diffusion.py

```diff
@@ -0,0 +1,174 @@
+
+.. math::
+    \int_\\Omega \\psi \\, p_t \\, dv =
+    \\alpha \int_{\\partial \\Omega} \\mathbf{n} \\cdot ( \\psi \\nabla p ) \\, da
+    - \\alpha \int_\\Omega (\\nabla \\psi ) \\cdot (\\nabla p ) \\, dv
+    - \int_\\Omega (\\psi \\, \mathbf{u} ) \\cdot \\nabla p dv
+    + \\psi \\, q
+
+Since the flux on the faces is zero at the boundaries, we can eliminate
+the first term on the right-hand side. By evaluating the inner products
+according to the finite volume approach we obtain:
+
+.. math::
+    \\mathbf{\\psi^T M_c p_t} = - \\alpha \\mathbf{\\psi^T G^T M_f G \\, p}
+    - \\mathbf{\\psi^T M_c \\tilde{A}_{fc}} \\textrm{diag} ( \\mathbf{u} ) \\mathbf{G \\, p}
+    + \\mathbf{\\psi^T M_c s}
+
+where :math:`\\mathbf{\\psi}`, :math:`\\mathbf{p}` and :math:`\\mathbf{p_t}`
+live at cell centers and :math:`\\mathbf{u}` lives on faces. :math:`\\mathbf{G}`
+is the discrete gradient operators. :math:`\\mathbf{M_c}` and
+:math:`\\mathbf{M_f}` are the cell center and face inner product matricies,
+respectively. :math:`\\mathbf{\\tilde{A}_{cf}}` averages the x, y and z
+contributions of :math:`\\mathbf{u} \\cdot \\nabla p` to cell centers and sums them.
+
+By eliminating :math:`\\psi^T` and multiplying both sides by
+:math:`\\mathbf{M_c^{-1}}` we obtain:
+
+.. math::
+    \\mathbf{p_t} = - \\mathbf{M} \\mathbf{p} + \\mathbf{s}
+
+where
+
+.. math::
+    \\mathbf{M} = \\alpha \\mathbf{M_c^{-1} G^T M_f G}
+    + \\mathbf{A_{fc}} \\textrm{diag}(\\mathbf{u}) \\mathbf{G}
+
+For the example, we will discretize in time using backward Euler. This results
+in the following system which must be solve at every time step :math:`k`.
+Where :math:`\\Delta t` is the step size:
+
+.. math::
+    \\big [ \\mathbf{I} + \\Delta t \\, \\mathbf{M} \\big ] \\mathbf{p}^{k+1} =
+    \\mathbf{p}^k + \\Delta t \\, \\mathbf{s}
+    
+
+
+"""
+
+###################################################
+#
+# Import Packages
+# ---------------
+#
+# Here we import the packages required for this tutorial.
+#
+
+from discretize import TensorMesh
+from pymatsolver import SolverLU
+import matplotlib.pyplot as plt
+import numpy as np
+from discretize.utils.matutils import sdiag
+
+###############################################
+#
+# Solving the Problem
+# -------------------
+#
+
+# Create a tensor mesh
+h = np.ones(75)
+mesh = TensorMesh([h, h], 'CC')
+
+# Define diffusivity constant
+a = 25
+
+# Define velocity vector u = [ux, uy] = [1, -1] on faces
+ux = 10*np.ones((mesh.nFx))
+uy = - 10*np.ones((mesh.nFy))
+u = np.r_[ux, uy]
+
+# Define source term diag(v)*s
+xycc = mesh.gridCC
+k1 = (xycc[:, 0]==-25) & (xycc[:, 1]==25)   # source at (-25, 25)
+k2 = (xycc[:, 0]==25) & (xycc[:, 1]==-25)   # sink at (25, -25)
+
+s = np.zeros(mesh.nC)
+s[k1] = 10
+s[k2] = 10
+
+# Define system M
+Afc = mesh.dim*mesh.aveF2CC
+Mc_inv = sdiag(1/mesh.vol)
+Mf = mesh.getFaceInnerProduct()
+
+mesh.setCellGradBC(['neumann', 'neumann'])  # Set Neumann BC
```

Comment:
Now we are mixing finite difference and finite volume ... (actually I am not sure this is working..)

Can we rather use nodal discretization? If you need to set up a Neumann boundary condition?
So, if you define `u` on edge and `p` on nodes. You are good to go. Need to use `mesh.nodalGrad`.

Reply:
1. The natural boundary conditions for this problem should be no flux entering or leaving. This is more straightforward to implement if the gradient is centers to faces.
2. Even if I did this the way you are suggesting,  psi and diag(u)*G*p live at different places. As a result, we cannot just use one of the inner product matricies we have.
3. I also how doubt this example is working properly.

We should discuss this in more detail.

Reply:
:wave: @dccowan,@sgkang, have you had a chance to connect on this?

Reply:
I have looked at the way I am solving the problem and made some changes. I believe it is much better now but it should undergo a final pass by @sgkang

Reply:
Looks good for now, and working.